diff --git a/pkg/executor/adapter.go b/pkg/executor/adapter.go
index e5ea851a10..5982aff910 100644
--- a/pkg/executor/adapter.go
+++ b/pkg/executor/adapter.go
@@ -1441,6 +1441,11 @@ func (a *ExecStmt) handlePessimisticLockError(ctx context.Context, lockErr error
 
 	a.resetPhaseDurations()
 
+	if err = resetCTEStorageMap(a.Ctx); err != nil {
+		return nil, err
+	}
+	a.Ctx.GetSessionVars().StmtCtx.CTEStorageMap = map[int]*CTEStorages{}
+
 	a.inheritContextFromExecuteStmt()
 	e, err := a.buildExecutor(ctx)
 	if err != nil {
